<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>
<body>
  <script>
    // 构造函数
    function Axios(config){
      this.config = config;
      this.interceptors = {
        request : new InterceptorManager(),
        response : new InterceptorManager(),
      }
    }

    Axios.prototype.request = function(config){ // 作整合成一个大数组，请求的回调放前面，响应拦截的回调放后面。然后通过 promise 链条的方式去执行里面的回调。
      // 创建一个 promise 对象
      let promise = Promise.resolve(config);
      // 创建一个数组
      const chains = [dispatchRequest, undefined];
      // 处理拦截器
      // 处理请求拦截器 将请求拦截器的回调压入到 chains 的前面
      this.interceptors.request.handlers.forEach(item => {
        chains.unshift(item.fulfilled, item.rejected);
      })
      // 处理响应拦截器
      this.interceptors.response.handlers.forEach(item => {
        chains.push(item.fulfilled, item.rejected);
      })

      // 遍历 (为什么不会覆盖？本质是通过 promise 链条的方式逐一往后执行。第一个因为 promise 声明就是成功的，所以肯定执行成功回调，所以 then 得到的 promise 也是成功的，所以执行下一个拦截器的成功的回调..然后执行dispatchRequest模拟请求获取结果。然后执行响应拦截器成功回调...chains中执行完后结束。返回最终经过响应拦截器处理后（预处理）的结果给用户)
      while(chains.length > 0){
        promise = promise.then(chains.shift(), chains.shift());
      }

      return promise;
    }

    // 这里做演示就不去调用适配器了，直接返回结果
    function dispatchRequest(config){
      return new Promise((resolve, reject) => {
        resolve({
          status: 200,
          statusText: 'OK',
        })
      })
    }

    // 创建实例
    let context = new Axios({});
    // 创建 axios 函数
    let axios = Axios.prototype.request.bind(context);
    // 将 context 属性 config interceptors 添加至 axios 函数对象身上
    Object.keys(context).forEach(key => {
      axios[key] = context[key];
    })

    console.dir(axios);

    // 拦截器管理器构造函数 (通过这两函数保存回调，用的时候在 request 里用，负责作整合)
    function InterceptorManager(){
      this.handlers = [];
    } 
    InterceptorManager.prototype.use = function(fulfilled, rejected){
      this.handlers.push({
        fulfilled,
        rejected
      })
    }


    // 执行顺序：2号请求拦截器 => 1号请求拦截器 => 1号响应拦截器 => 2号响应拦截器 => 用户回调函数
    // 添加请求拦截器
    axios.interceptors.request.use(function (config) { // 两个参数：成功回调、失败回调。 加入到了 request 身上的 handlers 数组内
      console.log('请求拦截器 1号 成功');
      return config;
    }, function (error) {
      console.log('请求拦截器 1号 失败')
      return Promise.reject(error);
    });

    axios.interceptors.request.use(function (config) { // 执行第二个请求拦截器后 request 的属性 handlers数组增加了第二个值。数组每个元素保存了对应拦截器的成功回调&失败回调
      console.log('请求拦截器 2号 成功');
      return config;
    }, function (error) {
      console.log('请求拦截器 2号 失败')
      return Promise.reject(error);
    });

    // 添加响应拦截器
    axios.interceptors.response.use(function (response) { // 到 axios 的时候，它对这个数组进行完善，将请求拦截器回调往原数组前面放，将响应拦截器回调往后放。最终形成了2号请求拦截器回调在1号前面，而响应拦截器回调顺序正常的局面。
      console.log('响应拦截器 1号 成功');
      return response;
    }, function (error) {
      console.log('响应拦截器 1号 失败')
      return Promise.reject(error);
    });

    axios.interceptors.response.use(function (response) {
      console.log('响应拦截器 2号 成功');
      return response;
    }, function (error) {
      console.log('响应拦截器 2号 失败')
      return Promise.reject(error);
    });

    console.dir(axios);
    axios({ // 本质是调用 request 函数。其内的 this 指向 context 这个 Axios 实例
      method: 'GET',
      url: 'http://localhost:3000/posts'
    }).then(response => {
      console.log(response);
    })

  </script>
</body>
</html>